File: solr/webapp/web/js/angular/controllers/documents.js
/*
 Licensed to the Apache Software Foundation (ASF) under one or more
 contributor license agreements.  See the NOTICE file distributed with
 this work for additional information regarding copyright ownership.
 The ASF licenses this file to You under the Apache License, Version 2.0
 (the "License"); you may not use this file except in compliance with
 the License.  You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
//helper for formatting JSON and others

var DOC_PLACEHOLDER = '<doc>\n' +
                '<field name="id">change.me</field>' +
                '<field name="title">change.me</field>' +
                '</doc>';

var ADD_PLACEHOLDER = '<add>\n' + DOC_PLACEHOLDER + '</add>\n';

solrAdminApp.controller('DocumentsController',
    function($scope, $rootScope, $routeParams, $location, Luke, Update, FileUpload, Constants) {
        $scope.resetMenu("documents", Constants.IS_COLLECTION_PAGE);

        $scope.refresh = function () {
            Luke.schema({core: $routeParams.core}, function(data) {
                //TODO: handle dynamic fields
                delete data.schema.fields._version_;
                $scope.fields = Object.keys(data.schema.fields);
            });
            $scope.document = "";
            $scope.handler = "/update";
            $scope.type = "json";
            $scope.commitWithin = 1000;
            $scope.overwrite = true;
            $scope.boost = "1.0";
        };

        $scope.refresh();

        $scope.changeDocumentType = function () {
            $scope.placeholder = "";
            if ($scope.type == 'json') {
                $scope.placeholder = '{"id":"change.me","title":"change.me"}';
            } else if ($scope.type == 'csv') {
                $scope.placeholder = "id,title\nchange.me,change.me";
            } else if ($scope.type == 'solr') {
                $scope.placeholder = ADD_PLACEHOLDER;
            } else if ($scope.type == 'xml') {
                $scope.placeholder = DOC_PLACEHOLDER;
            }
        };

        $scope.addWizardField = function () {
            if ($scope.document == "") $scope.document = "{}";
            var doc = JSON.parse($scope.document);
            doc[$scope.fieldName] = $scope.fieldData;
            $scope.document = JSON.stringify(doc, null, '\t');
            $scope.fieldData = "";
        };

        $scope.submit = function () {
            var contentType = "";
            var postData = "";
            var params = {};
            var doingFileUpload = false;

            if ($scope.handler[0] == '/') {
                params.handler = $scope.handler.substring(1);
            } else {
                params.handler = 'update';
                params.qt = $scope.handler;
            }

            params.commitWithin = $scope.commitWithin;
            params.boost = $scope.boost;
            params.overwrite = $scope.overwrite;
            params.core = $routeParams.core;
            params.wt = "json";

            if ($scope.type == "json" || $scope.type == "wizard") {
                postData = "[" + $scope.document + "]";
                contentType = "json";
            } else if ($scope.type == "csv") {
                postData = $scope.document;
                contentType = "csv";
            } else if ($scope.type == "xml") {
                postData = "<add>" + $scope.document + "</add>";
                contentType = "xml";
            } else if ($scope.type == "upload") {
                doingFileUpload = true;
                params.raw = $scope.literalParams;
            } else if ($scope.type == "solr") {
                postData = $scope.document;
                if (postData[0] == "<") {
                    contentType = "xml";
                } else if (postData[0] == "{" || postData[0] == '[') {
                    contentType = "json";
                } else {
                    alert("Cannot identify content type")
                }
            }
            if (!doingFileUpload) {
                var callback = function (success) {
                  $scope.responseStatus = "success";
                  delete success.$promise;
                  delete success.$resolved;
                  $scope.response = JSON.stringify(success, null, '  ');
                };
                var failure = function (failure) {
                    $scope.responseStatus = failure;
                };
                if (contentType == "json") {
                  Update.postJson(params, postData, callback, failure);
                } else if (contentType == "xml") {
                  Update.postXml(params, postData, callback, failure);
                } else if (contentType == "csv") {
                  Update.postCsv(params, postData, callback, failure);
                }
            } else {
                var file = $scope.fileUpload;
                console.log('file is ' + JSON.stringify(file));
                var uploadUrl = "/fileUpload";
                FileUpload.upload(params, $scope.fileUpload, function (success) {
                    $scope.responseStatus = "success";
                    $scope.response = JSON.stringify(success, null, '  ');
                }, function (failure) {
                    $scope.responseStatus = "failure";
                    $scope.response = JSON.stringify(failure, null, '  ');
                });
            }
        }
    });

File: lucene/luke/src/java/org/apache/lucene/luke/models/documents/Documents.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.lucene.luke.models.documents;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import org.apache.lucene.index.Term;
import org.apache.lucene.luke.models.LukeException;

/** A dedicated interface for Luke's Documents tab. */
public interface Documents {

  /** Returns one greater than the largest possible document number. */
  int getMaxDoc();

  /** Returns field names in this index. */
  Collection<String> getFieldNames();

  /**
   * Returns true if the document with the specified <code>docid</code> is not deleted, otherwise
   * false.
   *
   * @param docid - document id
   */
  boolean isLive(int docid);

  /**
   * Returns the list of field information and field data for the specified document.
   *
   * @param docid - document id
   * @throws LukeException - if an internal error occurs when accessing index
   */
  List<DocumentField> getDocumentFields(int docid);

  /** Returns the current target field name. */
  String getCurrentField();

  /**
   * Returns the first indexed term in the specified field. Empty Optional instance is returned if
   * no terms are available for the field.
   *
   * @param field - field name
   * @throws LukeException - if an internal error occurs when accessing index
   */
  Optional<Term> firstTerm(String field);

  /**
   * Increments the terms iterator and returns the next indexed term for the target field. Empty
   * Optional instance is returned if the terms iterator has not been positioned yet, or has been
   * exhausted.
   *
   * @return next term, if exists, or empty
   * @throws LukeException - if an internal error occurs when accessing index
   */
  Optional<Term> nextTerm();

  /**
   * Seeks to the specified term, if it exists, or to the next (ceiling) term. Returns the term that
   * was found. Empty Optional instance is returned if the terms iterator has not been positioned
   * yet, or has been exhausted.
   *
   * @param termText - term to seek
   * @return found term, if exists, or empty
   * @throws LukeException - if an internal error occurs when accessing index
   */
  Optional<Term> seekTerm(String termText);

  /**
   * Returns the first document id (posting) associated with the current term. Empty Optional
   * instance is returned if the terms iterator has not been positioned yet, or the postings
   * iterator has been exhausted.
   *
   * @return document id, if exists, or empty
   * @throws LukeException - if an internal error occurs when accessing index
   */
  Optional<Integer> firstTermDoc();

  /**
   * Increments the postings iterator and returns the next document id (posting) for the current
   * term. Empty Optional instance is returned if the terms iterator has not been positioned yet, or
   * the postings iterator has been exhausted.
   *
   * @return document id, if exists, or empty
   * @throws LukeException - if an internal error occurs when accessing index
   */
  Optional<Integer> nextTermDoc();

  /**
   * Returns the list of the position information for the current posting.
   *
   * @throws LukeException - if an internal error occurs when accessing index
   */
  List<TermPosting> getTermPositions();

  /**
   * Returns the document frequency for the current term (the number of documents containing the
   * current term.) Empty Optional instance is returned if the terms iterator has not been
   * positioned yet.
   *
   * @throws LukeException - if an internal error occurs when accessing index
   */
  Optional<Integer> getDocFreq();

  /**
   * Returns the term vectors for the specified field in the specified document. If no term vector
   * is available for the field, empty list is returned.
   *
   * @param docid - document id
   * @param field - field name
   * @return list of term vector elements
   * @throws LukeException - if an internal error occurs when accessing index
   */
  List<TermVectorEntry> getTermVectors(int docid, String field);

  /**
   * Returns the doc values for the specified field in the specified document. Empty Optional
   * instance is returned if no doc values is available for the field.
   *
   * @param docid - document id
   * @param field - field name
   * @return doc values, if exists, or empty
   * @throws LukeException - if an internal error occurs when accessing index
   */
  Optional<DocValues> getDocValues(int docid, String field);
}
